### PR TITLE
Use typed configuration for Rundeck configs.

### PR DESCRIPTION
**Pull request by sjrd218**
This PR is designed to provide a starting place for adding and using Rundeck configs in a type safe way. The underlying mechanism uses Spring's configuration mapping feature.

Please give feedback regarding the technical solution and structuring of the typed configs.

### CODE REVIEW

File: rundeckapp/grails-app/services/rundeck/services/ConfigurationService.groovy [hunk 1]
```diff
@@ -212,4 +241,17 @@ class ConfigurationService {
             return defval
         }
     }
+
+    @Override
+    void afterPropertiesSet() throws Exception {
+        appCfg = grailsApplication?.config?.rundeck
+    }
+
+    Map<String,String> deprecatedKeyTrx = [
```
Comment (gschueler) [inline rundeckapp/grails-app/services/rundeck/services/ConfigurationService.groovy]: maybe this map could be defined put into resources.groovy, or a config of its own, making it easier to extend later if more things come up, unless these are the only values we might need to translate

File: rundeckapp/src/main/groovy/org/rundeck/app/services/EnhancedNodeService.groovy [hunk 2]
```diff
@@ -52,7 +52,7 @@ class EnhancedNodeService
 
     @Override
     void afterPropertiesSet() throws Exception {
-        enabled = featureService.featurePresent('enhanced-nodes', false)
+        enabled = featureService.featurePresent('enhancedNodes', false)
```
Comment (gschueler) [inline rundeckapp/src/main/groovy/org/rundeck/app/services/EnhancedNodeService.groovy]: we might want a Constants class for feature names

File: rundeckapp/grails-app/services/rundeck/services/ConfigurationService.groovy (hunk 1 above)
Comment (sjrd218) [reply rundeckapp/grails-app/services/rundeck/services/ConfigurationService.groovy]: As long as future configs are added to the typed configuration it should prevent devs from using config names that can't be mapped to Java property names.

Since these keys already exist, there needed to be some way of falling back from the new name to get the value out of the old if it exists. 

I hope this map doesn't need any additions.

File: rundeckapp/src/main/groovy/org/rundeck/app/services/EnhancedNodeService.groovy (hunk 2 above)
Comment (sjrd218) [reply rundeckapp/src/main/groovy/org/rundeck/app/services/EnhancedNodeService.groovy]: I guess we could have a Features class that has all of the current features. In the pro side we could extend the class like ProFeatures and add the pro only features. What do you think?

File: rundeckapp/build.gradle [hunk 3]
```diff
@@ -29,6 +29,13 @@ apply plugin:"com.energizedwork.idea-project-components"
 apply plugin:"asset-pipeline"
 apply plugin:"org.grails.grails-gsp"
 
+dependencyManagement {
+    imports {
+        mavenBom "org.grails:grails-bom:$grailsVersion"
+        mavenBom 'org.springframework.cloud:spring-cloud-starter-parent:Edgware.SR6'
```
Comment (ahormazabal) [inline rundeckapp/build.gradle]: Curious that you had to add the grails and spring-cloud boms to the build. This shouldn't be necessary to include spring dependencies. did you run with some issues?

File: rundeckapp/grails-app/services/rundeck/services/ConfigurationService.groovy (hunk 1 above)
Comment (ahormazabal) [reply rundeckapp/grails-app/services/rundeck/services/ConfigurationService.groovy]: I would keep this mapping close to where properties are defined, to make easier for developers to find it. Probably this class is fine, or RundeckConfigBase. Also maybe changing this to an enum class could allow you to both define a constant, and map both old an new property keys to keep track of this legacy properties.

File: rundeckapp/src/main/groovy/org/rundeck/app/services/EnhancedNodeService.groovy (hunk 2 above)
Comment (ahormazabal) [reply rundeckapp/src/main/groovy/org/rundeck/app/services/EnhancedNodeService.groovy]: Yes i agree. probably an enum class would allow us to define both constants and associated mappings when needed.

File: rundeckapp/build.gradle (hunk 3 above)
Comment (sjrd218) [reply rundeckapp/build.gradle]: I had to add this when I had my dev environment configured with the new enterprise reloadable config plugin. The way our gradle project is set up caused gradle to be unable to find the dependencies until I added this section on the OSS side. We could probably leave it out for the OSS build, but if anyone tries to use the pro config plugin in their dev env it will fail if this section isn't present. Ideas?

File: rundeckapp/grails-app/views/framework/projectNodeSources.gsp [hunk 4]
```diff
@@ -82,7 +82,7 @@
                         </a>
                       </li>
 
-                    <feature:enabled name="enhanced-nodes">
+                    <feature:enabled name="enhancedNodes">
```
Comment (gschueler) [inline rundeckapp/grails-app/views/framework/projectNodeSources.gsp:85]: maybe the feature:enabled tag lib can accept Feature enum type attribute e.g. `<feature:enabled feature="${Features.ENHANCED_NODES}">`

File: rundeckapp/grails-app/conf/application.groovy [hunk 5]
```diff
@@ -116,6 +113,7 @@ grails.plugin.springsecurity.interceptUrlMap = [
         [pattern: '/user/loggedout', access: ['permitAll']],
         [pattern: '/feed/**',        access: ['permitAll']],
         [pattern: '/api/**',         access: ['permitAll']],
+        [pattern: '/health',         access: ['permitAll']],
```
Comment (gschueler) [inline rundeckapp/grails-app/conf/application.groovy:116]: this seems unrelated to typed configs, why was this added?

File: rundeckapp/build.gradle [hunk 6]
```diff
@@ -102,8 +102,8 @@ dependencies {
     compile "org.springframework.boot:spring-boot-autoconfigure"
 
     compile "org.grails:grails-core"
-    //compile "org.springframework.boot:spring-boot-starter-actuator"
-    compile ("org.springframework.boot:spring-boot-starter-jetty"){
+    compile "org.springframework.boot:spring-boot-starter-actuator"
```
Comment (gschueler) [inline rundeckapp/build.gradle:105]: is the actuator necessary?

File: rundeckapp/grails-app/conf/application.groovy (hunk 5 above)
Comment (sjrd218) [reply rundeckapp/grails-app/conf/application.groovy:116]: This is added for the Pro side of this feature. When Rundeck registers itself with Consul, Consul likes to have a health check endpoint so it can test the apps that have registered and make sure they're still alive. The /heath endpoint is the standard endpoint it hits.

File: rundeckapp/build.gradle (hunk 6 above)
Comment (sjrd218) [reply rundeckapp/build.gradle:105]: The actuator was added to expose the health endpoint and also to add a refresh endpoint. After further research I'm not sure we really need this. I was using it as a quick way to expose the /health check endpoint, but we can do that ourselves without the other features that actuator also adds.

File: rundeckapp/grails-app/views/framework/projectNodeSources.gsp (hunk 4 above)
Comment (sjrd218) [reply rundeckapp/grails-app/views/framework/projectNodeSources.gsp:85]: I think we may need to reevaluate the Features enum. It's not a very flexible way to handle referencing features. If we are looking at adding and removing features at runtime I think we need something more flexible.

Comment (gschueler) [reply rundeckapp/grails-app/views/framework/projectNodeSources.gsp:85]: hmm, perhaps a Feature bean type? the specific feature name can be a constant string in the implementation class
